doc/reference/javascript_reference.rst

```diff
@@ -0,0 +1,1806 @@
+---------------------
+
+The custom event widgets is a more advanced system, which mimic the DOM events
+API.  Whenever an event is triggered, it will 'bubble up' the component tree,
+until it reach the root widget, or is stopped.
+
+- *trigger_up*: this is the method that will create a small *OdooEvent* and
+  dispatch it in the component tree.  Note that it will start with the component
+  that triggered the event
+- *custom_events*: this is the equivalent of the *event* dictionary, but for
+  odoo events.
+
+The OdooEvent class is very simple.  It has three public attributes: *target*
+(the widget that triggered the event), *name* (the event name) and *data* (the
+payload).  It also has 2 methods: *stopPropagation* and *is_stopped*.
+
+The previous example can be updated to use the custom event system:
+
+.. code-block:: javascript
+
+    var Widget = require('web.Widget');
+    var Counter = require('myModule.Counter');
+
+    var MyWidget = Widget.extend({
+        custom_events: {
+            valuechange: '_onValueChange'
+        },
+        start: function () {
+            this.counter = new Counter(this);
+            return this.counter.appendTo(this.$el);
+        },
+        _onValueChange: function(event) {
+            // do something with event.data.val
+        },
+    });
+
+    // in Counter widget, we need to call the trigger_up method:
+
+    ... this.trigger_up('valuechange', {value: someValue});
+
+
+Registries
+===========
+
+A common need in the Odoo ecosystem is to extend/change the behaviour of the
+base system from the outside (by installing an application, i.e. a different
+module).  For example, one may need to add a new widget type in some views.  In
+that case, and many others, the usual process is to create the desired component,
+then add it to a registry (registering step), to make the rest of the web client
+aware of its existence.
+
+There are a few registries available in the system:
+
+- field registry (exported by 'web.field_registry'). The field registry contains
+    all field widgets known to the web client.  Whenever a view (typically, form,
+    or list/kanban) needs a field widget, this is where it will look. A typical
+    use case look like this:
+
+    .. code-block:: javascript
+
+        var fieldRegistry = require('web.field_registry');
+
+        var FieldPad = ...;
+
+        fieldRegistry.add('pad', FieldPad);
+
+    Note that each value should be a subclass of *AbstractField*
+
+- view registry: this registry contains all JS views known to the web client
+    (and in particular, the view manager).  Each value of this registry should
+    be a subclass of *AbstractView*
+
+- action registry: we keep track of all client actions in this registry.  This
+    is where the action manager look up whenever it needs to create a client
+    action.  In version 11, each value should simply be a subclass of *Widget*.
+    However, in version 12, the values are required to be *AbstractAction*.
```

Comment:
As this is the documentation fort v11, I wouldn't mention this here, but I would do a commit in master to change it.